### ISSUE TITLE
Javascript files (e.g., shiny.min.js) are not being served for Rmd files in subdirectory

### ISSUE THREAD

**Issue by ralmond** [rstudio/shiny-server#429]
I'm running Shiny Server version 1.5.12.933  under Node.js v10.15.3 under RHEL and I'm running as a proxy from Apache 2.4.6 under RHEL 7.5

This setup works fine if the Rmd file is in the top level directory, but has difficulties if the Rmd file is in a subdirectory.  Thus the link:

https://pluto.coe.fsu.edu/rdemos/NormalCalculator.Rmd

works as expected, but 

https://pluto.coe.fsu.edu/rdemos/IntroStats/Correlation.Rmd

does not load.  Looking at the web console the problem is the files `json2-min.js`, `shiny.min.js`, `jquery.min.js`, and `rmd_loader.js` (in the `shared` and `rmd_resources`) subdirectories return 404 errors when the browser tries to load them.  The URL's look properly formatted, but are looking inside of the `rdemos/IntroStats/shared` and not `redmos/shared` directory (if that makes a difference).

As I'm handling the redirect from https to shiny with the apache proxy, I tried adding specific redirects, as shown with the commented lines below, but that is not working.

Here is my `shiny.conf` redirection file:
```
####
### RGA
### This sets up the redirect from /rdemos directory to the Shiny
### server.
<VirtualHost _default_:80>
  ##RGA -- not sure if these lines are necessary.
<Proxy *>
   Allow from localhost
</Proxy>


  RewriteEngine on
  RewriteCond %{HTTP:Upgrade} =websocket
#  RewriteRule /rdemos/.*/shared/(.*) ws://localhost:3838/shared/$1 [P,L]
#  RewriteRule /rdemos/.*/rmd_resources/(.*) ws://localhost:3838/rmd_resources/$1 [P,L]
  RewriteRule /rdemos/(.*) ws://localhost:3838/$1 [P,L]
  RewriteCond %{HTTP:Upgrade} !=websocket
  RewriteRule /rdemos/(.*) http://localhost:3838/$1 [P,L]
#  RewriteRule /rdemos/.*/shared/(.*) http://localhost:3838/shared/$1 [P,L]
#  RewriteRule /rdemos/.*/rmd_resources/(.*) http://localhost:3838/rmd_resources/$1 [P,L]
  ProxyPass /rdemos/ http://localhost:3838/
  ProxyPassReverse /rdemos/ http://localhost:3838/

  Header edit Location ^/ /rdemos/
  ProxyRequests Off

  RedirectMatch permanent ^/rdemos$ /rdemos/


</VirtualHost>
``` 
The problem persists whether or not the commented lines are uncommented.

The problem appeared suddenly several months ago, and I simply have not been able to get it working, and I've received no reply from my posts on the RStudio message board.  Even a pointer to where in the code to look would be helpful.

**Comment by jcheng5** [comment]
To narrow it down, does the /IntroStats/ version work if you bypass the proxy entirely and hit port 3838 instead?

**Comment by ralmond** [comment]
I'm getting exactly the same issue when I go directly without the proxy severer:
```
Loading failed for the <script> with source “http://localhost:3838/IntroStats/shared/json2-min.js”.
```

**Comment by ralmond** [comment]
Is is possible that the issue is with my shiny-server.conf?

Here it is:
```
# Define a server that listens on port 3838
server {
  listen 3838;

  # Define a location at the base URL
  location / {

    # Host the directory of Shiny Apps stored in this directory
    site_dir /www1/shiny-server;

    # Log all Shiny output to files in this directory
    log_dir /var/log/shiny-server;


    # When a user visits the base URL rather than a particular application,
    # an index of the applications available in this directory will be shown.
    directory_index on;

    ## RGA want to run as apache not root
    run_as shiny;

    log_file_mode 0644;

    #preserve_logs true;

     ## Define a location at the IntroStats directory
  #   location /IntroStats {

  #     ## When a user visits the base URL rather than a particular application,
  #     ## an index of the applications available in this directory will be shown
.
  #    directory_index on;

  #    ## This is an attempt to fix a bug where javascript utils are not redirect
ing properly.	
  #    location /IntroStats/shared {
  #    	redirect "http://localhost:3838/shared" 301;
  #    } 
  #    location /IntroStats/rmd_resources {
  #    	redirect "http://localhost:3838/rmd_resources" 301;
  #    } 
	  	
  }		
  # # Define a location at the EDA directory
  # location /EDA {

  #   # Host the directory of Shiny Apps stored in this directory
  #   site_dir /www1/shiny-server/EDA;

  #   # Log all Shiny output to files in this directory
  #   log_dir /var/log/shiny-server;

  #   # When a user visits the base URL rather than a particular application,
  #   # an index of the applications available in this directory will be shown.
  #   directory_index on;

  #   ## RGA want to run as apache not root
  #   run_as shiny;

  # }

  # # Define a location at the Bayesian directory
  # location /Bayesian {

  #   # Host the directory of Shiny Apps stored in this directory
  #   site_dir /www1/shiny-server/Bayesian;

  #   # Log all Shiny output to files in this directory
  #   log_dir /var/log/shiny-server;

  #   # When a user visits the base URL rather than a particular application,
  #   # an index of the applications available in this directory will be shown.
  #   directory_index on;

  #   ## RGA want to run as apache not root
  #   run_as shiny;

  # }

  # # Define a location at the Bayesian directory
  # location /RIntro {

  #   # Host the directory of Shiny Apps stored in this directory
  #   site_dir /www1/shiny-server/RIntro;

  #   # Log all Shiny output to files in this directory
  #   log_dir /var/log/shiny-server;

  #   # When a user visits the base URL rather than a particular application,
  #   # an index of the applications available in this directory will be shown.
  #   directory_index on;

  #   ## RGA want to run as apache not root
  #   run_as shiny;

  # }


}
```
The commented out parts represent previous attempts at trying to get the subdirectory working, as well as to turn on directory indexing for the subdirectories.

**Comment by jcheng5** [comment]
And the Correlation.Rmd file is located in `/www1/shiny-server/rdemos/IntroStats/`?

**Comment by jcheng5** [comment]
Can you also do a `ls -l /www1/shiny-server/rdemos`? Certain files in parent directories can cause subdirectories to not be considered app directories (for reasons I can explain if this turns out to be the problem)

**Comment by ralmond** [comment]
The relevant directory is `/www1/shiny-server`.  The rdemos is used by Apache to route applications to the shiny server and is take out of the path.

The results of the ls command are:
```
[ralmond@pluto shiny-server]$ ls -al
total 156
drwxrwxr-x. 11 shiny shiny 4096 Jan 16 14:02 .
drwxr-xr-x. 13 root  root   199 Nov 12 13:02 ..
drwxrwxr-x.  2 shiny shiny 4096 Sep 11 17:42 Bayesian
-rw-r--r--.  1 shiny shiny 6831 Jan 16 14:00 Correlation.Rmd
drwxr-xr-x.  2 shiny shiny   53 Nov 27  2018 data
drwxrwxr-x.  2 shiny shiny  106 Oct 22 13:53 EDA
-rw-r--r--.  1 shiny shiny 2393 Jan 16 14:00 Geyser.Rmd
-rw-rw-r--.  1 shiny shiny  679 Jan  9 17:30 header.html
-rw-r--r--.  1 shiny shiny 1781 Nov 27  2018 index.html
drwxrwxr-x.  5 shiny shiny 4096 Jan 16 12:59 IntroStats
-rw-r--r--.  1 shiny shiny 2885 Oct 22 13:53 MatrixesAndDataFrames.Rmd
-rw-r--r--.  1 shiny shiny 4632 Nov  5 16:59 NormalCalculator.Rmd
lrwxrwxrwx.  1 root  root    27 Jan 16 13:57 NormalParams.Rmd -> IntroStats/NormalParams.Rmd
-rw-rw-r--.  1 shiny shiny 2335 Sep 11 17:42 RatersStan.Rmd
-rw-rw-r--.  1 shiny shiny  205 Jan 24  2019 rgroup-shiny.Rproj
drwxrwxr-x.  2 shiny shiny  156 Sep 11 17:42 RIntro
drwxrwxr-x.  4 shiny shiny   30 Nov 20  2018 sample-apps
-rw-rw-r--.  1 shiny shiny 4887 Nov  7 18:40 Studenttcalculator.Rmd
drwxrwxr-x.  4 shiny shiny   75 Jan 16 12:59 .svn
drwxrwxr-x.  2 shiny shiny   61 Sep 11 17:42 Talks
lrwxrwxrwx.  1 shiny shiny   23 Jan 16 14:02 Untitled.Rmd -> IntroStats/Untitled.Rmd
-r-sr-sr-x.  1 shiny shiny 8448 Nov 27  2018 updateRshiny
drwxr-xr-x.  2 shiny shiny   19 Jan 31  2019 UpdateSVN
lrwxrwxrwx.  1 root  root    23 Jan 16 13:57 Z-scores.Rmd -> IntroStats/Z-scores.Rmd
[ralmond@pluto shiny-server]$ 
```
(I've cleaned out a few files that copied from IntroStats directory to the top level directory as a workaround.)  The updateRshiny is a script for updating the site from a subversion repro.  The contents of the UpdateSVN directory is an app.R file which is also not currently working.

The contents of the IntroStats directory is:
```
[ralmond@pluto shiny-server]$ ls -al IntroStats/
total 128
drwxrwxr-x.  5 shiny shiny 4096 Jan 16 12:59 .
drwxrwxr-x. 11 shiny shiny 4096 Jan 16 14:02 ..
-rw-rw-r--.  1 shiny shiny 5225 Sep 11 17:42 ConfidenceInterval.Rmd
-rw-rw-r--.  1 shiny shiny 5954 Jan 31  2019 CorrelationOutliers.Rmd
-rw-r--r--.  1 shiny shiny 6831 Jan 14 11:02 Correlation.Rmd
-rw-rw-r--.  1 shiny shiny 5580 Jan 26  2019 EffectSize.Rmd
-rw-r--r--.  1 shiny shiny 2393 Sep 24 13:20 Geyser.Rmd
-rw-r--r--.  1 shiny shiny 3253 Sep 24 13:20 index.html
-rw-r--r--.  1 shiny shiny  559 Sep 24 13:20 index.Rhtml
-rw-rw-r--.  1 shiny shiny 4120 Sep 11 17:42 KurtosisBoxplots.Rmd
-rw-rw-r--.  1 shiny shiny 4547 Sep 11 17:42 KurtosisPractice.Rmd
-rw-rw-r--.  1 shiny shiny 4966 Sep 11 17:42 KurtosisQQ.Rmd
-rw-rw-r--.  1 shiny shiny 4425 Sep 11 17:42 LawOfLargeNumbers.Rmd
drwxrwxr-x.  2 shiny shiny   19 Feb 19  2019 NormalParameters
-rw-r--r--.  1 shiny shiny 2755 Oct 22 13:53 NormalParams.Rmd
drwxr-xr-x.  2 shiny shiny   49 Jan  9 17:36 rmd_resources
drwxr-xr-x.  2 shiny shiny   84 Jan  9 17:34 shared
-rw-rw-r--.  1 shiny shiny  525 Jan 31  2019 sign_turn_left.png
-rw-rw-r--.  1 shiny shiny 4103 Sep 11 17:42 SkewnessBoxplot.Rmd
-rw-rw-r--.  1 shiny shiny 4544 Sep 11 17:42 SkewnessPractice.Rmd
-rw-rw-r--.  1 shiny shiny 4952 Sep 11 17:42 SkewnessQQ.Rmd
-rw-rw-r--.  1 shiny shiny 3178 Jan 31  2019 StandardDeviations.Rmd
-rw-rw-r--.  1 shiny shiny  835 Jan 16 12:59 Untitled.Rmd
-rw-r--r--.  1 shiny shiny 2576 Jan 14 11:02 Z-scores.Rmd
```
The `shared` and `rmd_resources` were attempts to work around the problem that did not work, and were added after teh problem developed.

**Comment by jcheng5** [comment]
So, the problem is as I suspected; if a directory contains .Rmd files, then that directory is considered an app directory. Once a directory is considered an app directory, then its child directories can't be considered app directories, nor will they send directory listings to the browser. The reason for this is that applications are made up of potentially private code and data; anything within reach of an app.R or .Rmd (i.e. in their directory or below) are considered private, i.e. it's up to the app.R or .Rmd to serve up those contents if they want to.

You can serve up app subdirectories, but you have to explicitly tell Shiny Server to do so. You almost had it correct:

```r
  location / {

    # Host the directory of Shiny Apps stored in this directory
    site_dir /www1/shiny-server;

    # Log all Shiny output to files in this directory
    log_dir /var/log/shiny-server;


    # When a user visits the base URL rather than a particular application,
    # an index of the applications available in this directory will be shown.
    directory_index on;

    ## RGA want to run as apache not root
    run_as shiny;

    log_file_mode 0644;

    #preserve_logs true;

    ## Define a location at the IntroStats directory
    location /IntroStats {
      site_dir /www1/shiny-server/IntroStats;
    }
  }
```

The part that was missing was the `site_dir` directive under `location /IntroStats`.

(Also, there's no need to have `shared` directories anywhere, those will be automatically served up by the Shiny package.)

**Comment by jcheng5** [comment]
(I'm pretty confident the above fixes the issue, so I'm closing it, but feel free to reopen if it doesn't solve it.)

**Comment by ralmond** [comment]
Thanks for the tips on writing the config files.  I found the documentation confusing on this issue.  I had previously tired both with and without the `site_dir` directive, and I found that this did not resolve the issue.  I just retried it and that alone was not enough to resolve the problem.

Moving all of the Rmd files out of the top level directory did fix the problem.  

I think the issue should be reopened to fix the documentation if nothing else.
